### PR TITLE
BlockEditor driven by shared ObservableStore

### LINKED ISSUE #987: Load and persist block editor from actual data
Part of #214

### PR DESCRIPTION
**Pull request by gordonbrander**
Fixes #987. 

Depends on https://github.com/subconsciousnetwork/ObservableStore/pull/43

This PR implements basic loading and saving of real data from database, including blocks and related notes. I've kept the scope to *just* saving and loading, and there is more work to be done wrt transcludes and loading states.

## Approach

- New changes to ObservableStore allow creating custom Update types for models. https://github.com/subconsciousnetwork/ObservableStore/pull/43
- Also introduces an `updates` publisher on stores
- Custom update type lets us include a change description along with state changes, supplying extra information that the ViewController might need to perform these changes.
- Root UIViewController subscribes to `updates` publisher and performs updates to controller based on state transactions.
- ObservableStore instance is stored as a `@StateObject` on the enclosing SwiftUIView and passed down as an `@ObservedObject` to the UIViewControllerRepresentable.

The result is that we can share the same store between SwiftUI and UIKit land, as well as the same environment and side effects management.

## Screenshots

<table>
  <tr>
    <td>
      
    </td>
    <td>
      
    </td>
  </tr>
</table>

### CODE REVIEW

**Comment by gordonbrander** [comment]
Depends on https://github.com/subconsciousnetwork/ObservableStore/pull/41

File: xcode/Subconscious/Shared/Parsers/SubtextAttributedStringRenderer.swift [hunk 1]
```diff
@@ -283,39 +283,22 @@ struct SubtextAttributedStringRenderer {
...
-        attributedString.addAttribute(
-            .foregroundColor,
-            value: UIColor(Color.primary),
-            range: baseNSRange
-        )
+        let baseNSRange = NSRange(location: 0, length: attributedString.length)
+        
+        // Set default styles for entire string.
+        // Clears all previous attributes.
+        attributedString.setAttributes(attributes, range: baseNSRange)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Parsers/SubtextAttributedStringRenderer.swift:301]: Wow, much nicer.

File: xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorModel.swift [hunk 2]
```diff
@@ -7,35 +7,1336 @@
...
+            return block
+        })
+        
+        return Update(
+            state: model,
+            change: .reconfigureCollectionItem(indexPath)
+        )
+    }
+
+    // TODO: re-implement
```
Comment (bfollington) [inline xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorModel.swift:956]: I'm assuming these TODOs are out of scope for this PR since they're related to bulk selection.

File: xcode/Subconscious/Shared/UIKit/BlockEditor/Blocks/BlocksModel.swift [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+//
+
+import Foundation
+
+extension BlockEditor {
+    /// A list of blocks making up an editable document
+    struct BlocksModel: Hashable, CustomStringConvertible {
+        var description: String {
+            // TODO: serialize to nice Subtext whitespace
+            blocks.compactMap(\.text).joined(separator: "\n\n")
```
Comment (bfollington) [inline xcode/Subconscious/Shared/UIKit/BlockEditor/Blocks/BlocksModel.swift:15]: Is this out of scope?

File: xcode/Subconscious/Shared/UIKit/BlockEditor/Common/SubtextTextView.swift [hunk 4]
```diff
@@ -12,31 +12,42 @@ extension UIView {
...
+            // UITextView has a bug where a text view that has been created
+            // with an empty string does not correctly render attributes until
+            // given a non-empty string. This results in a jump triggered by
+            // a change in intrinsic content size when the attributes are
+            // first applied.
+            //
+            // After being rendered with a non-empty string, it works properly.
+            // 2023-11-22 Gordon
+            self.text = " "
+            self.text = ""
```
Comment (bfollington) [inline xcode/Subconscious/Shared/UIKit/BlockEditor/Common/SubtextTextView.swift:48]: image

**Review by bfollington — APPROVED** [review]
Nice work, this is a massive improvement to the existing code. Excited to try the editor out.

File: xcode/Subconscious/Shared/UIKit/BlockEditor/Blocks/BlocksModel.swift (hunk 3 above)
Comment (gordonbrander) [reply xcode/Subconscious/Shared/UIKit/BlockEditor/Blocks/BlocksModel.swift:15]: Done

File: xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorModel.swift (hunk 2 above)
Comment (gordonbrander) [reply xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorModel.swift]: Yes, out of scope. The selection mode wasn't working properly. Removed commented-out code, and added issues to TODOs.
